kolibri/core/discovery/utils/network/search.py

```diff
@@ -0,0 +1,175 @@
+import atexit
+import json
+import logging
+import socket
+import time
+from contextlib import closing
+
+from zeroconf import get_all_addresses
+from zeroconf import NonUniqueNameException
+from zeroconf import ServiceInfo
+from zeroconf import USE_IP_OF_OUTGOING_INTERFACE
+from zeroconf import Zeroconf
+
+from kolibri.core.auth.models import Facility
+from kolibri.core.content.models import ChannelMetadata
+
+logger = logging.getLogger(__name__)
+
+SERVICE_TYPE = "Kolibri._sub._http._tcp.local."
+LOCAL_DOMAIN = "kolibri.local"
+
+ZEROCONF_STATE = {"zeroconf": None, "listener": None, "service": None}
+
+
+def _id_from_name(name):
+    assert name.endswith(SERVICE_TYPE), (
+        "Invalid service name; must end with '%s'" % SERVICE_TYPE
+    )
+    return name.replace(SERVICE_TYPE, "").strip(".")
+
+
+def _is_port_open(host, port, timeout=1):
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
```

Comment:
Add a local cache so that this check isn't done too frequently, but so that the check is done:
- immediately following discovery
- after a while (~ 20 seconds?)

Reply:
I'd suggest having it be a "shared" cache (not just local memory), using Django's cache backend, because it might be different web service processes handling the sequential queries.